### PR TITLE
The real obstructed method :-)

### PR DESCRIPTION
**Pull request by RaykPlatzek**
We had to push this new branch up as we got problems with merging two branches locally.
@mscoutermarsh can you please have a final review of the method (we managed to change it so that we do not need to test any nil conditions anymore)

### CODE REVIEW

**Comment by RaykPlatzek** [comment]
will resolve the merge conflicts later today

**Comment by RaykPlatzek** [comment]
Merge conflicts resolved...ready to be pushed to master once reviewed

File: app/models/piece.rb [hunk 1]
```diff
@@ -1,8 +1,147 @@
...
+      elsif test == true && index == 2
+        destination_tests << diagonal_obstruction_up_right?(destination_y)
+      elsif test == true && index == 3
+        destination_tests << diagonal_obstruction_down_left?(destination_y)
+      elsif test == true && index == 4
+        destination_tests << diagonal_obstruction_up_left?(destination_y)
+      elsif test == true && index == 5
+        destination_tests << diagonal_obstruction_down_right?(destination_y)
+      end
+    end
```
Comment (mscoutermarsh) [inline app/models/piece.rb]: This is getting really close. You can simplify this code more to make it easier to follow.

Try a pattern like this

```Ruby
def obstructed?(destination_x, destination_y)
  raise 'Error: Invalid Input' if destination_y > 7 || destination_x > 7 || destination_y < 0 || destination_x < 0

  # if is a certain type of move, check if that direction is obstructed.

  if diagonal_down_and_right_move?(destination_x, destination_y)
    return diagonal_obstruction_down_right?(destination_y)
  end

  #... test other directions

  if horizontal_move?(destination_y)
    return horizontal_obstruction_left?(destination_x) && horizontal_obstruction_right?(destination_x)
  end

  raise 'Error: Invalid Input'
end
```

When thinking about a method. Try to make it read like a paragraph. Another a developer should be able to come along and read top-to-bottom and mostly understand what it's doing.

File: app/models/piece.rb [hunk 2]
```diff
@@ -1,8 +1,147 @@
...
+    return true if destination_tests.include?(true)
+    false
+  end
+
+  def horizontal_move?(destination_y)
```
Comment (mscoutermarsh) [inline app/models/piece.rb]: All of these methods can be `private`

(Add `private` above this method and all methods below it will then be private)

http://culttt.com/2015/06/03/the-difference-between-public-protected-and-private-methods-in-ruby/

**Comment by mscoutermarsh** [comment]
I think this PR can be closed right? The other one was merged.

**Comment by roderickjoseph** [comment]
@mike Yes, this one can be closed.
